data/src/file_upload_session.rs

```diff
@@ -119,92 +104,128 @@ impl FileUploadSession {
-            self.xorb_uploader.register_new_cas_block(new_cas_data).await?;
+        #[cfg(debug_assertions)]
+        {
+            let current_session_data = self.current_session_data.lock().await;
+            debug_assert_le!(current_session_data.num_bytes(), *MAX_XORB_BYTES);
+            debug_assert_le!(current_session_data.num_chunks(), *MAX_XORB_CHUNKS);
         }
 
-        let total_bytes_trans = self.xorb_uploader.flush().await?;
+        // Now, aggregate the new dedup metrics.
+        self.deduplication_metrics.lock().await.merge_in(dedup_metrics);
+
+        Ok(())
+    }
+
+    /// Process the aggregated data, uploading the data as a xorb and registering the files
+    async fn process_aggregated_data_as_xorb(&self, data_agg: DataAggregator) -> Result<()> {
+        let (xorb, new_files) = data_agg.finalize();
+        debug_assert_le!(xorb.num_bytes(), *MAX_XORB_BYTES);
+        debug_assert_le!(xorb.data.len(), *MAX_XORB_CHUNKS);
 
-        // flush accumulated memory shard.
-        self.shard_manager.flush().await?;
+        self.xorb_uploader.register_new_xorb_for_upload(xorb).await?;
 
-        if !self.dry_run {
-            self.upload_shards().await?;
+        for fi in new_files {
+            self.shard_interface.add_file_reconstruction_info(fi).await?;
         }
 
-        Ok(total_bytes_trans)
+        Ok(())
     }
 
-    async fn upload_shards(&self) -> Result<()> {
-        // First, get all the shards prepared and load them.
-        let merged_shards_jh = self.remote_shards.merge_shards()?;
+    /// Finalize with  
+    async fn finalize_impl(self: Arc<Self>, return_files: bool) -> Result<(DeduplicationMetrics, Vec<MDBFileInfo>)> {
+        // This should be used as if it's consuming the class, as it effectively empties all the states.
+        debug_assert_eq!(Arc::strong_count(&self), 1);
 
-        // Get a list of all the merged shards in order to upload them.
-        let merged_shards = merged_shards_jh.await??;
+        // Register the remaining xorbs for upload.
+        let data_agg = take(&mut *self.current_session_data.lock().await);
+        self.process_aggregated_data_as_xorb(data_agg).await?;
 
-        // Now, these need to be sent to the remote.
-        self.remote_shards.upload_and_register_shards(merged_shards).await?;
+        // Now, make sure all the remaining xorbs are uploaded.
+        let mut metrics = take(&mut *self.deduplication_metrics.lock().await);
 
-        // Finally, we can move all the mdb shards from the session directory, which is used
-        // by the upload_shard task, to the cache.
-        self.remote_shards.move_session_shards_to_local_cache().await?;
+        metrics.xorb_bytes_uploaded = self.xorb_uploader.finalize().await?;
 
-        Ok(())
+        let all_file_info = {
+            if return_files {
+                self.shard_interface.session_file_info_list().await?
+            } else {
+                Vec::new()
+            }
+        };
```

Comment:
syntax option here, I personally prefer the suggested change, but go with your preference/what is consistent here
```suggestion
        let all_file_info = if return_files {
            self.shard_interface.session_file_info_list().await?
        } else {
            Vec::new()
        };
```